File: exp-15/q2-kmp-string-matching/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func computeLPS(p string) []int {
	m := len(p)
	lps := make([]int, m)
	length, i := 0, 1
	for i < m {
		if p[i] == p[length] {
			length++
			lps[i] = length
			i++
		} else if length != 0 {
			length = lps[length-1]
		} else {
			lps[i] = 0
			i++
		}
	}
	return lps
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	text := scanner.Text()
	scanner.Scan()
	pattern := scanner.Text()
	n, m := len(text), len(pattern)
	lps := computeLPS(pattern)
	i, j := 0, 0
	found := false
	for i < n {
		if pattern[j] == text[i] {
			i++
			j++
		}
		if j == m {
			fmt.Printf("Pattern found at index %d\n", i-j)
			found = true
			j = lps[j-1]
		} else if i < n && pattern[j] != text[i] {
			if j != 0 {
				j = lps[j-1]
			} else {
				i++
			}
		}
	}
	if !found {
		fmt.Println("not found")
	}
}

File: exp-15/q2-kmp-string-matching/solution.c
#include <stdio.h>
#include <string.h>

void trim(char *s)
{
    s[strcspn(s, "\r\n")] = '\0';
}

void computeLPS(char *p, int m, int lps[])
{
    lps[0] = 0;
    int len = 0, i = 1;
    while (i < m)
    {
        if (p[i] == p[len])
        {
            len++;
            lps[i] = len;
            i++;
        }
        else if (len != 0)
        {
            len = lps[len - 1];
        }
        else
        {
            lps[i] = 0;
            i++;
        }
    }
}

int main()
{
    char text[1000], pattern[1000];
    fgets(text, sizeof(text), stdin);
    fgets(pattern, sizeof(pattern), stdin);
    trim(text);
    trim(pattern);
    int n = strlen(text), m = strlen(pattern);
    int lps[1000];
    computeLPS(pattern, m, lps);
    int i = 0, j = 0, found = 0;
    while (i < n)
    {
        if (pattern[j] == text[i])
        {
            i++;
            j++;
        }
        if (j == m)
        {
            printf("Pattern found at index %d\n", i - j);
            found = 1;
            j = lps[j - 1];
        }
        else if (i < n && pattern[j] != text[i])
        {
            if (j != 0)
                j = lps[j - 1];
            else
                i++;
        }
    }
    if (!found)
        printf("not found\n");
    return 0;
}
